### PR TITLE
feat: singletons + enable creation of preflight objects in static

### LINKED ISSUE #513: Builtin support for singleton resources
### Summary

Support for singleton resource patterns built into the Wing language

### Feature Spec

See alternative proposal [here](https://github.com/winglang/wing/issues/513#issuecomment-1729747103).

### Use Cases

One-of resources used throughout your entire application, like a database, or a logger, etc.

### Implementation Notes

Based on how `Stack.of` or `App.of` is implemented in the AWS CDK (https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/core/lib/stack.ts) or how `Logger.of` is implemented in the Wing SDK.

### Component

Language Design

### LINKED ISSUE #2583: Cannot instantiate classes from preflight static methods
### I tried this:

```js
bring cloud;

class Foo {
  static makeFunc(): cloud.Function {
    return new cloud.Function(inflight (s: str): str => {
      return "hello world!";
    });
  }
}

let bar = Foo.makeFunc();

test "it works" {
  let result = bar.invoke("");
  assert(result == "hello world!");
}
```

### This happened:

```
$ wing compile -t sim statics.w       
Cannot read properties of undefined (reading 'root')
```

### I expected this:

It compiles and the test passes

### Is there a workaround?

Change the method from a static method to a regular method

### Component

Compiler

### Wing Version

0.17.7

### Wing Console Version

_No response_

### Node.js Version

18.14.1

### Platform(s)

MacOS

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #4246: Unable to pass `IConstruct` as an `in` scope
### I tried this:

```js
class C1 { }
class C2 {
  init() {
    new C1() in std.Node.of(this).root;
  }
}
```

### This happened:

```
Expected scope to be a preflight object, instead found "IConstruct"
```

### I expected this:

`in` should accept `IConstruct`. The use case is interaction with the construct tree.

### Is there a workaround?

Yes, but it's horribly ugly and requires an extern cast:

```js
bring "constructs" as constructs;

class Scope {
  extern "./util.js" static from(c: constructs.IConstruct): Scope;
}

class C1 { }
class C2 {
  init() {
    new C1() in Scope.from(std.Node.of(this).root);
  }
}
```

`util.js`

```js
exports.from = x => x;
```

### Component

Compiler

### Wing Version

_No response_

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #4987: Unable to instantiate JSII objects in static methods
### I tried this:

```js
class Foo {
  static makeBar(scope: std.IResource) {
    return new jsii.module.MyClass() in scope;
  }
}
```

### This happened:

Even though I am explicitly stating the scope (there’s no implicit scope in static methods at the moment), this fails because the JS output is `this.node.root.new("Bar", Bar, …`

### I expected this:

To work

### Is there a workaround?

_No response_

### Anything else?

We need to revisit the whole dependency injection mechanism I think

### Wing Version

_No response_

### Node.js Version

_No response_

### Platform(s)

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #5015: can't test app with custom providers and more than one test in tf-aws
### I tried this:

```
bring cloud;
bring util;
bring "@cdktf/provider-docker" as docker;
bring "@cdktf/provider-null" as nullProvider;
bring "@cdktf/provider-random" as randomProvider;

pub class MyCustomProviderResource {
  pub imageName: str;
  new() {
    new docker.provider.DockerProvider();
    new nullProvider.provider.NullProvider();
    new randomProvider.provider.RandomProvider();

    let randomString = new randomProvider.stringResource.StringResource(
      length: 10
    );

    let image = new docker.image.Image(
      name: "foo/bar",
    );

    let resource = new nullProvider.resource.Resource(triggers: { "changed": util.nanoid() });
    this.imageName = image.name;
  }
}


new MyCustomProviderResource();

test "foo" {
  log("hello");
}

test "bar" {
  log("hello");
}
```

run `wing test -t tf-aws main.w`

### This happened:

```
Initializing the backend...
Terraform encountered problems during initialisation, including problems
with the configuration, described below.

The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.
╷
│ Error: Duplicate provider configuration
│
│   on main.tf.json line 92, in provider:
│   92:     "docker": [
│
│ A default (non-aliased) provider configuration for "docker" was already
│ given at main.tf.json:92,15-16. If multiple configurations are required,
│ set the "alias" argument for alternative configurations.
╵

╷
│ Error: Duplicate provider configuration
│
│   on main.tf.json line 96, in provider:
│   96:     "null": [
│
│ A default (non-aliased) provider configuration for "null" was already given
│ at main.tf.json:96,13-14. If multiple configurations are required, set the
│ "alias" argument for alternative configurations.
╵

╷
│ Error: Duplicate provider configuration
│
│   on main.tf.json line 100, in provider:
│  100:     "random": [
│
│ A default (non-aliased) provider configuration for "random" was already
│ given at main.tf.json:100,15-16. If multiple configurations are required,
│ set the "alias" argument for alternative configurations.
```

excerpt from the synthesized config

```
 "provider": {
    "aws": [
      {}
    ],
    "docker": [
      {},
      {}
    ],
    "null": [
      {},
      {}
    ],
    "random": [
      {},
      {}
    ],
    "dnsimple": [
      {}
    ]
  },
```

### I expected this:

no error

### Is there a workaround?

fix the synthesized config in postprocessing

### Anything else?

also tried workarounds mentioned here https://github.com/winglang/wing/issues/513

here's a repo for reproduction https://github.com/skorfmann/wing-test-providers

### Wing Version

0.48.8

### Node.js Version

18.7

### Platform(s)

MacOS, Linux

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### PR DESCRIPTION
**Pull request by eladb**
## Support the singleton pattern

Resolves #513 by making it easy to access the root app from every node through `Node.of(this).app` or `Node.of(this).root`. Both return the same root app, even if it's not the absolute root of the tree in case the app is part of a surrounding testing tree.

> In the future, this is supposed to become `nodeof(this).app`.

Also add `tryFindChild(id)` to the app to make it easier to implement app-level singletons:

```js
let uid = "unique-id";
let root = Node.of(this).root;
return unsafeCast(root.tryFindChild(uid)) ?? new MySingleton() as uid in root;
```

Added a doc example with this information.

Fixes #5015

## Enable creation of preflight classes in static methods

Fix #2583 by requiring `in scope` when creating preflight objects in static methods:

```js
class Foo {
  static maker(scope: std.IResource): Foo {
    return new Foo() in scope;
  }
}
```

Also, emit an error if there is no explicit scope.

```js
class Foo {
  static maker(): Foo {
    return new Foo();
//         ^^^^^^^ Cannot instantiate without an explicit scope
  }
}
```

This is also working for JSII classes and resolves #4987. It required a modification on how we emit `this.node.new()` to use the explicit scope instead of `this` in this case.

Resolves #4246 by allowing any `IConstruct` to be used within `in scope`.

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [x] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingsdk/src/std/resource.ts [hunk 1]
```diff
@@ -197,24 +197,7 @@ export abstract class Resource extends Construct implements IResource {
...
-    }
-
-    if (appScope) {
-      return appScope.newAbstract(fqn, scope, id, ...props);
-    } else {
-      throw new Error(
-        "This is an abstract class and can only be instantiated in the scope of a Wing App"
-      );
-    }
+    return App.of(scope).newAbstract(fqn, scope, id, ...props);
```
[the remark this answers is not in the document]
Comment (eladb) [reply libs/wingsdk/src/std/resource.ts:200]: thanks, got it.

in wing it's impossible to reach this case afaik - you can't define a resource outside of an app context, so i don't think the error message is that important.

are you okay with just changing the test expectation to `Cannot find root app`?

[the remark this answers is not in the document]
Comment (hasanaburayyan) [reply libs/wingsdk/src/std/resource.ts:200]: I think you are right, Im fine with the change.

File: libs/wingsdk/src/std/node.ts [hunk 2]
```diff
@@ -326,6 +345,21 @@ export class Node {
   public lock() {
     this._constructsNode.lock();
   }
+
+  /**
+   * Returns the root app.
+   */
+  public findApp(scope: IConstruct): IApp {
+    if ((scope as IApp)[APP_SYMBOL]) {
```
Comment (hasanaburayyan) [inline libs/wingsdk/src/std/node.ts:353]: 🔥

File: docs/docs/07-examples/80-singletons.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+keywords: [singleton]
+---
+
+Sometimes, it makes sense for a resource to be defined only once per application. This example shows
+a pattern for how to implement singletons in Wing.
+
+As an example, say we want to maintain a central bucket that can be accessed from any part of the
+application:
+
+```js
```
Comment (hasanaburayyan) [inline docs/docs/07-examples/80-singletons.md:13]: Id imagine people will want to play with this in playground
```suggestion
```js playground
```

**Review by hasanaburayyan — APPROVED** [review]
LGTM

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 4]
```diff
@@ -2556,6 +2578,16 @@ error: Expected type to be \\"preflight (inflight (str): str): void\\", but got
    |                  ^^^^^^^^ Expected type to be \\"preflight (inflight (str): str): void\\", but got \\"preflight (inner_fn: preflight (preflight (str): str): void): void\\" instead
 
 
+error: Cannot instantiate preflight class \\"$Closure3\\" in a static method without an explicit scope
```
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:2581]: Nope!

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 5]
```diff
@@ -2974,11 +3006,11 @@ exports[`scope_and_id.test.w 1`] = `
   |                         ^ Expected type to be \\"str\\", but got \\"num\\" instead
 
 
-error: Expected scope to be a preflight object, instead found \\"num\\"
+error: Expected scope to be a construct, instead found \\"num\\"
```
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:3009]: nope

File: tools/hangar/__snapshots__/test_corpus/valid/bring_cdk8s.test.w_compile_tf-aws.md [hunk 6]
```diff
@@ -47,7 +47,7 @@ class $Root extends $stdlib.std.Resource {
     super($scope, $id);
     const app = this.node.root.new("cdk8s.App", cdk8s.App, );
     const chart = this.node.root.new("cdk8s.Chart", cdk8s.Chart, this, "cdk8s.Chart");
-    const deploy = this.node.root.new("cdk8s-plus-27.Deployment", kplus.Deployment, chart, "kplus.Deployment");
+    const deploy = ($scope => $scope.node.root.new("cdk8s-plus-27.Deployment", kplus.Deployment, chart, "kplus.Deployment"))(chart);
```
Comment (eladb) [inline tools/hangar/__snapshots__/test_corpus/valid/bring_cdk8s.test.w_compile_tf-aws.md:50]: nope

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 7]
```diff
@@ -2339,6 +2339,24 @@ error: Expected type to be \\"MutMap<str>\\", but got \\"MutMap<num>\\" instead
 
  
  
+Tests 1 failed (1)
+Test Files 1 failed (1)
+Duration <DURATION>"
+`;
+
+exports[`new_in_static_without_scope.test.w 1`] = `
```
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:2347]: nope

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 8]
```diff
@@ -2556,6 +2578,13 @@ error: Expected type to be \\"preflight (inflight (str): str): void\\", but got
    |                  ^^^^^^^^ Expected type to be \\"preflight (inflight (str): str): void\\", but got \\"preflight (inner_fn: preflight (preflight (str): str): void): void\\" instead
 
 
+error: Unknown symbol \\"this\\"
```
Comment (eladb) [inline tools/hangar/__snapshots__/invalid.ts.snap:2581]: nope

Comment (eladb) [reply tools/hangar/__snapshots__/invalid.ts.snap:2581]: https://github.com/winglang/wing/issues/5034

**Pull request merged**